docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,372 @@
+func (dt DenomTrace) IBCDenom() string {
+  if dt.Trace != "" {
+    return fmt.Sprintf("ibc/%s", dt.Hash())
+  }
+  return dt.BaseDenom
+}
+```
+
+In order to trim the denomination trace prefix when sending/receiving fungible tokens, the `RemovePrefix` function is provided.
+
+> NOTE: the prefix addition must be done on the client side.
+
+```golang
+// RemovePrefix trims the first portID/channelID pair from the trace info. If the trace is already empty it will perform a no-op. If the trace is incorrectly constructed or doesn't have separators it will return an error.
+func (dt *DenomTrace) RemovePrefix() error {
+  if dt.Trace == "" {
+    return nil
+  }
+
+  traceSplit := strings.SplitN(dt.Trace, "/", 3)
+
+  var err error
+  switch {
+  case len(traceSplit) == 0, traceSplit[0] == dt.Trace:
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must contain '/' separators", dt.Trace)
+  case len(traceSplit) == 1:
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must come in pairs of '{portID}/channelID}'", dt.Trace)
+  case len(traceSplit) == 2:
+    dt.Trace = ""
+  case len(traceSplit) == 3:
+    dt.Trace = traceSplit[2]
+  }
+
+  if err != nil {
+    return err
+  }
+
+  return nil
+}
+```
+
+### `x/ibc-transfer` Changes
+
+In order to retreive the trace information from an IBC denomination, a lookup table needs to be
+added to the `ibc-transfer` module. These values need to also be persisted between upgrades, meaning
+that a new `[]DenomTrace` `GenesisState` field state needs to be added to the module:
+
+```golang
+// GetDenom retreives the full identifiers trace and base denomination from the store.
+func (k Keeper) GetDenomTrace(ctx Context, denomTraceHash []byte) (DenomTrace, bool) {
+  store := ctx.KVStore(k.storeKey)
+  bz := store.Get(types.KeyDenomTrace(traceHash))
+  if bz == nil {
+    return &DenomTrace, false
+  }
+
+  var denomTrace DenomTrace
+  k.cdc.MustUnmarshalBinaryBare(bz, &denomTrace)
+  return denomTrace, true
+}
+
+// HasTrace checks if a the key with the given trace hash exists on the store.
+func (k Keeper) HasDenomTrace(ctx Context, denomTraceHash []byte)  bool {
+  store := ctx.KVStore(k.storeKey)
+  return store.Has(types.KeyTrace(denomTraceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, denomTrace DenomTrace) {
+  store := ctx.KVStore(k.storeKey)
+  bz := k.cdc.MustMarshalBinaryBare(&denomTrace)
+  store.Set(types.KeyTrace(denomTrace.Hash()), bz)
+}
+```
+
+The problem with this approach is that when a token is received for the first time, the full trace
+info will need to be passed in order to construct the hash and set it to the mapping on the store.
+To mitigate this a new `Trace` field needs to be added to the `FungibleTokenPacketData`:
+
+```protobuf
+message FungibleTokenPacketData {
+  // the tokens to be transferred
+  repeated cosmos.Coin amount = 1 [
```

Comment:
@cwgoes shouldn't we extract out denom and amount of the token in this packet data into string and uint64? Otherwise wouldn't every implementation of `ics20` need to use `sdk.Coin`? 

I assumed this reasoning was why the sender and receiver are strings and not `sdk.AccAddress`

Reply:
This is a good point. Is there anything in `cosmos.Coin` other than denom & amount? Still, it might be cleaner to just have denom & amount as separate fields per ICS 20